clang/include/clang/CIR/Dialect/Passes.h

```diff
@@ -41,6 +41,8 @@ std::unique_ptr<Pass> createGotoSolverPass();
 /// Create a pass to lower ABI-independent function definitions/calls.
 std::unique_ptr<Pass> createCallConvLoweringPass();
 
+std::unique_ptr<Pass> createAddrSpaceLoweringPass();
```

Comment:
Why we need a separate pass to do address space lowering, I don't see any good reason why it shouldn't be done directly in CIRGen.

Reply:
I don't think it can be done in CIRGen immediately. Did you mean the TypeConverter of DirectToLLVM? I also prefer that way and already implemented it previously. If we use TypeConverter instead, we can also get rid of `recursivelyReplaceElementsIn` and `replaceVersion` (the handler should still be fixed though, can be another patch). It just bother a bit `ConvertCIRToLLVMPass::runOnOperation` to maintain the states required by LowerModule. If it's okay, let me update the changes.

Reply:
> It just bother a bit ConvertCIRToLLVMPass::runOnOperation to maintain the states required by LowerModule

Can you elaborate a bit what would be needed specifically? Would it require a hacky solution or just some extra work?

Reply:
Sure. This is the patch looks like:
https://gist.github.com/seven-mile/c57ddfb96e180af1609d429e1e66f67b

Apart from messing up `ConvertCIRToLLVMPass::runOnOperation`, it doesn't look too hacky IMO.

Reply:
Yes, perhaps I've been too insistent on using an ideal method to call TargetLoweringInfo. We could opt for an interim solution (passing AS map by value) and refine it as it becomes more accessible.

A feasible approach would be:

```cpp
// All CIR AS maps have static lifetime, it's safe to ref it
auto getCIRASMap = [&]() -> mlir::cir::AddressSpaceAttr::map_t const & {
  // If the triple is not present, e.g. CIR modules parsed from text, we
  // cannot init LowerModule properly.
  assert(!::cir::MissingFeatures::makeTripleAlwaysPresent());
  // Here we will use a default AS map to ignore all AS stuff.
  static const mlir::cir::AddressSpaceAttr::map_t defaultASMap{0};
  if (module->hasAttr("cir.triple")) {
    mlir::PatternRewriter rewriter{&getContext()};
    auto lowerModule = mlir::cir::createLowerModule(module, rewriter);
    return lowerModule.getTargetLoweringInfo().getCIRAddrSpaceMap();
  } else {
    return defaultASMap;
  }
};
prepareTypeConverter(converter, dataLayout, getCIRASMap());
```